File: BDA-Java/src/main/java/utn/frc/bda/semana03/introOOP/polimorfismoEstVsDin/Poli.cpp
/******************************************************************************

                              Online C++ Compiler.
               Code, Compile, Run and Debug C++ program online.
Write your code in this editor and press "Run" button to compile and execute it.

*******************************************************************************/

#include <iostream>

class Base {
    public:
        void f() {
            std::cout<<"base\n";
        }
};

class Derivada {
    public:
        void f() {
            std::cout<<"derivada\n";
        }
};

int main()
{
    std::cout<<"Hello World\n";

    Base *b = new Base;
    Derivada *d = new Derivada;

    b->f();                 // base
    d->f();                 // derivada

    ((Base*) d)->f();       // base (derivada en Java)

    b = (Base*) d;
    b->f();                 // base (derivada en Java)
    ((Derivada*) b)->f();   // en Java NO se puede

    return 0;
}

File: BDA-Java/src/main/java/utn/frc/bda/semana03/introOOP/polimorfismoEstVsDin/Poli.pas
{

                            Online Pascal Compiler.
                Code, Compile, Run and Debug Pascal program online.
Write your code in this editor and press "Run" button to execute it.

}

{$mode objfpc}
{$m+}

program Hello;

type

    Base = class
        public
            procedure Proc;
    end;

    Derivada = class(Base)
        public
            procedure Proc;
    end;

    PBase = ^Base;
    PDerivada = ^Derivada;


procedure Base.Proc;
begin
    writeln('base');
end;

procedure Derivada.Proc;
begin
    writeln('derivada');
end;


var
    b: Base;
    d: Derivada;
    pb: PBase;
    pd: PDerivada;

begin
  writeln ('Hello World');

  b := Base.Create;
  d := Derivada.Create;

  b.Proc;
  d.Proc;
  b := d;
  b.Proc;
  (b as Derivada).Proc;

end.